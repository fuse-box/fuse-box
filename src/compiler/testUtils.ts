import { generate } from './generator/generator';
import { ASTNode } from './interfaces/AST';
import { ImportType } from './interfaces/ImportType';
import { ITransformer } from './interfaces/ITransformer';
import { ITransformerRequireStatementCollection } from './interfaces/ITransformerRequireStatements';
import { parseTypeScript } from './parser';
import { createGlobalContext } from './program/GlobalContext';
import { transpileModule } from './program/transpileModule';
import { GlobalContextTransformer } from './transformers/GlobalContextTransformer';

function cleanupForTest(node) {
  delete node.loc;
  delete node.raw;
  delete node.range;
}
export function initCommonTransform(props: {
  jsx?: boolean;
  code: string;
  transformers: Array<ITransformer>;
  props?: any;
}) {
  const requireStatementCollection: ITransformerRequireStatementCollection = [];
  function onRequireCallExpression(importType: ImportType, statement: ASTNode) {
    // making sure we have haven't emitted the same property twice
    if (!statement['emitted']) {
      Object.defineProperty(statement, 'emitted', { enumerable: false, value: true });
      cleanupForTest(statement.arguments[0]);
      cleanupForTest(statement);
      cleanupForTest(statement.callee);

      requireStatementCollection.push({ importType, statement });
    }
  }

  const ast = parseTypeScript(props.code, { jsx: props.jsx });

  // const ast = buntis.parseTSModule(props.code, {
  //   directives: true,
  //   jsx: props.jsx,
  //   next: true,
  //   loc: true,
  // });
  const tranformers = [GlobalContextTransformer().commonVisitors(props.props)];
  for (const t of props.transformers) {
    if (t.commonVisitors) {
      tranformers.push(t.commonVisitors({ ...props.props, onRequireCallExpression }));
    }
  }
  transpileModule({
    ast: ast as ASTNode,
    globalContext: createGlobalContext(),
    transformers: tranformers,
  });
  const res = generate(ast, {});

  return { code: res, requireStatementCollection };
}
