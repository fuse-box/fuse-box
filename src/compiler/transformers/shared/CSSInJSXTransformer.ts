import { path2RegexPattern } from '../../../utils/utils';
import { IVisit } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { GlobalContext } from '../../program/GlobalContext';

export interface CSSInJSXTransformerOptions {
  autoInject?: boolean;
  autoLabel?: boolean;
  cssPropOptimization?: boolean;
  emotionCoreAlias?: string;
  jsxFactory?: string;
  labelFormat?: string;
  module?: any;
  sourceMap?: boolean;
  test?: RegExp | string;
}

const labelMapping = {
  '[dirname]': '',
  '[filename]': '',
  '[local]': '',
};

/**
 * 1. compare minifiers styled-components and emotion
 * 2. don't reinvent the wheel..
 * 3. find a place for these helper functions..
 */
// From https://github.com/styled-components/babel-plugin-styled-components/blob/master/src/minify/index.js#L58
// Counts occurences of substr inside str
const countOccurences = (str, substr) => str.split(substr).length - 1;
const compressSymbols = code =>
  code.split(/(\s*[;:{},]\s*)/g).reduce((str, fragment, index) => {
    // Even-indices are non-symbol fragments
    if (index % 2 === 0) {
      return str + fragment;
    }

    // Only manipulate symbols outside of strings
    if (countOccurences(str, "'") % 2 === 0 && countOccurences(str, '"') % 2 === 0) {
      return str + fragment.trim();
    }

    return str + fragment;
  }, '');

// Super simple minifier.. doesn't cover any edge cases or side effects
const minify = (value: string): string => compressSymbols(value.replace(/[\n]\s*/g, ''));

/**
 * @todo
 * 1. expand the minify (way to simple :P)
 *
 * 2. Components as selectors
 * 3. Minification
 * 4. Sourcemaps
 * 5. Minify configurable?
 *
 * 5. Dead Code Elimination // not needed??
 */
export function CSSInJSXTransformer(options?: CSSInJSXTransformerOptions): ITransformer {
  const {
    autoInject = true,
    autoLabel = true,
    // cssPropOptimization = true,
    emotionCoreAlias = '@emotion/core',
    jsxFactory = 'jsx',
    labelFormat = '[dirname]--[local]',
    test = /\.(js|jsx|ts|tsx)$/,
    // sourceMap = true
  } = options;

  const testPathRegex = path2RegexPattern(test);
  const emotionLibraries = [[emotionCoreAlias, 'emotion'], ['@emotion/styled']];
  // const reactLibrary = 'react';

  function renderAutoLabel(): ASTNode {
    return {
      type: 'Literal',
      value: `label:${labelFormat
        .replace(/\[local\]|\[filename\]|\[dirname\]/gi, m => labelMapping[m])
        .replace(/\-\-$/, '')};`,
    };
  }

  return {
    target: { test: testPathRegex },
    commonVisitors: ({ ctx, module }) => {
      // Keep track of imported emotion functions.
      // Allows us to look for custom imports
      // import { css as emotionCss } from '@emotion/core'
      const importedEmotionFunctions = [];
      let needsInjection = true;

      // Check if we're executing an emotion call in this file
      function isEmotionCall(node: ASTNode, index: string): boolean {
        return (
          // css('styles')
          importedEmotionFunctions.indexOf(node[index].name) > -1 ||
          // styled('obj')('styles')
          importedEmotionFunctions.indexOf(node[index].callee && node[index].callee.name) > -1 ||
          // styled.div('div')
          importedEmotionFunctions.indexOf(node[index].object && node[index].object.name) > -1
        );
      }

      const compilerJsxFactory = ctx && ctx.compilerOptions.jsxFactory;

      // Process dirName and fileName only once per file
      const filePath = module.props.fuseBoxPath.replace(/\.([^.]+)$/, '');
      labelMapping['[dirname]'] = filePath.replace(/(\\|\/)/g, '-');
      labelMapping['[filename]'] = filePath.replace(/(.+)(\\|\/)(.+)$/, '$3');

      return {
        onEachNode: (visit: IVisit) => {
          const { node, parent } = visit;

          switch (node.type) {
            case 'CallExpression':
              // append the css class label
              if (autoLabel && !parent.callee && isEmotionCall(node, 'callee')) {
                labelMapping['[local]'] =
                  (parent.type === 'VariableDeclarator' && parent.id.name) ||
                  (parent.type === 'AssignmentExpression' && parent.left.property.name) ||
                  '';
                node.arguments.push(renderAutoLabel());
              }
              break;

            // case 'JSXElement':
            //   if (!cssPropOptimization) {
            //     break;
            //   }
            //   const {
            //     openingElement: { attributes },
            //   } = node;
            //   const attrLength = attributes.length;
            //   if (attrLength === 0) {
            //     break;
            //   }
            //   for (let i = 0; i < attrLength; i++) {
            //     let { name, type, value } = attributes[i]; // call 'attr' once
            //     if (type !== 'JSXAttribute' || name.name !== 'css') {
            //       continue;
            //     }
            //     if (value.expression.type === 'ObjectExpression' || value.expression.type === 'ArrayExpression') {
            //       // css prop optimization
            //       // console.log(name, value);
            //     }
            //   }
            //   break;

            case 'TaggedTemplateExpression':
              if (isEmotionCall(node, 'tag')) {
                let {
                  quasi: { expressions, quasis },
                  tag: callee,
                } = node;

                // Convert template strings to a literal and put the expressions back at it's position
                const styleProperties = [];
                const quasisLength = quasis.length;
                let i = 0;
                while (i < quasisLength) {
                  if (quasis[i].value.cooked) {
                    // We don't need minification in devMode!
                    styleProperties.push({
                      type: 'Literal',
                      value: !ctx.config.isProduction ? quasis[i].value.cooked : minify(quasis[i].value.cooked),
                    });
                  }
                  // Put the expressions back in the place where they belong
                  if (!quasis[i].tail && expressions.length > 0) {
                    styleProperties.push(expressions.shift());
                  }
                  i++;
                }

                // Replace this node with new shiny stuff
                return {
                  replaceWith: {
                    arguments: styleProperties.filter(Boolean),
                    callee,
                    type: 'CallExpression',
                  },
                };
              }
              break;
          }
        },
        onTopLevelTraverse: (visit: IVisit) => {
          const node = visit.node;
          const globalContext = visit.globalContext as GlobalContext;
          if (node.type === 'ImportDeclaration') {
            if (compilerJsxFactory === 'jsx') {
              // @todo:
              // fix this, always insert if it's jsx
            }

            if ([].concat(emotionLibraries[0], emotionLibraries[1]).indexOf(node.source.value) > -1) {
              const specifiersLength = node.specifiers.length;
              let i = 0;
              while (i < specifiersLength) {
                if (node.specifiers[i].imported && node.specifiers[i].imported.name === 'jsx') {
                  needsInjection = false;
                  // @todo:
                  // fix this, remove specifier if compilerJsxFactory === 'jsx'

                  // set the globalContext.jsxFactory for the JSXTransformer
                  globalContext.jsxFactory = node.specifiers[i].local.name;
                  // visit.globalContext.jsxFactory = node.specifiers[i].local.name;
                } else {
                  importedEmotionFunctions.push(node.specifiers[i].local.name);
                }
                i++;
              }

              if (autoInject && needsInjection && emotionLibraries[0].indexOf(node.source.value) > -1) {
                needsInjection = false;

                // set the globalContext.jsxFactory for the JSXTransformer
                if (node.source.value === emotionCoreAlias) {
                  // visit.globalContext.jsxFactory = jsxFactory;
                  globalContext.jsxFactory = jsxFactory;
                }
                node.specifiers.push({
                  imported: {
                    name: 'jsx',
                    type: 'Identifier',
                  },
                  local: {
                    name: jsxFactory,
                    type: 'Identifier',
                  },
                  type: 'ImportSpecifier',
                });
              }
            }
          }
        },
      };
    },
  };
}
