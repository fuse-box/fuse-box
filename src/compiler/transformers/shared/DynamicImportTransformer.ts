import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';
import { ImportType } from '../../interfaces/ImportType';
import { getDynamicImport } from '../importHelpers';

export function DynamicImportTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      return {
        onEachNode: (visit: IVisit): IVisitorMod => {
          const node = visit.node;

          const dynamicImport = getDynamicImport(node);
          if (dynamicImport) {
            if (dynamicImport.error) return;

            const requireCallExpression: ASTNode = {
              arguments: [
                {
                  type: 'Literal',
                  value: dynamicImport.source,
                },
              ],
              callee: {
                name: 'require',
                type: 'Identifier',
              },
              type: 'CallExpression',
            };

            if (props.onRequireCallExpression) {
              props.onRequireCallExpression(ImportType.DYNAMIC, requireCallExpression);
            }

            const callExpression: ASTNode = {
              arguments: [
                {
                  async: false,
                  body: requireCallExpression,
                  expression: true,
                  params: [],
                  type: 'ArrowFunctionExpression',
                },
              ],
              callee: {
                computed: false,
                object: {
                  arguments: [],
                  callee: {
                    computed: false,
                    object: {
                      name: 'Promise',
                      type: 'Identifier',
                    },
                    property: {
                      name: 'resolve',
                      type: 'Identifier',
                    },
                    type: 'MemberExpression',
                  },
                  type: 'CallExpression',
                },
                property: {
                  name: 'then',
                  type: 'Identifier',
                },
                type: 'MemberExpression',
              },
              type: 'CallExpression',
            };
            return {
              replaceWith: callExpression,
            };
          }

          return;
        },
      };
    },
  };
}
