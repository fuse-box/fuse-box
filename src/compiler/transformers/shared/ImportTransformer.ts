import * as path from 'path';
import { GlobalContext } from '../../program/GlobalContext';
import { ITransformer } from '../../program/transpileModule';
import { createRequireStatement } from '../../Visitor/helpers';
import { IVisit } from '../../Visitor/Visitor';
import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { ImportType } from '../../interfaces/ImportType';

// FIX SHORTHANDS
export function ImportTransformer(options?: ITransformerSharedOptions): ITransformer {
  options = options || {};
  return {
    onTopLevelTraverse: (visit: IVisit) => {
      const node = visit.node;
      const global = visit.globalContext as GlobalContext;

      if (node.type === 'ImportDeclaration') {
        // converts "./foo/bar.hello.js" to foo_bar_hello_js_1 (1:1 like typescript does)
        const variable = path.basename(node.source.value).replace(/\.|-/, '_') + '_' + global.getNextIndex();

        node.specifiers.forEach(specifier => {
          if (specifier.type === 'ImportSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: specifier.imported.name,
            };
          } else if (specifier.type === 'ImportDefaultSpecifier') {
            global.identifierReplacement[specifier.local.name] = {
              first: variable,
              second: 'default',
            };
          } else if (specifier.type === 'ImportNamespaceSpecifier') {
            // only if we have more than one specifier
            // for instance
            // i=mport MySuperClass, * as everything from "module-name";

            // in every other case like:
            // import * as everything from "module-name";
            // we don't need to do anything, since the variables should match

            global.identifierReplacement[specifier.local.name] = {
              first: variable,
            };

            //if (node.specifiers.length > 1) afterStatement = defineVariable(variable, specifier.local.name);
          }
        });

        return {
          onComplete: () => {
            const reqStatement = createRequireStatement(node.source.value, node.specifiers.length && variable);
            // when everything is finished we need to check if those variables have been used at all
            // they were all unused we need remove the require/import statement at all

            // assuming we have no specififers and this import HAS side effects
            // e.g import "./module"
            const parent = visit.parent;
            const property = visit.property;
            if (node.specifiers.length === 0) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.RAW_IMPORT, reqStatement.reqStatement);
                }
                parent[property].splice(index, 1, reqStatement.statement);
              }
              return;
            }

            let atLeastOneInUse = false;
            for (const specifier of node.specifiers) {
              const traced = global.identifierReplacement[specifier.local.name];
              if (traced && traced.inUse) {
                atLeastOneInUse = true;
                break; // we just need to know if we need to keep the node
              }
            }
            // doing a manual replace
            if (atLeastOneInUse) {
              const index = parent[property].indexOf(node);
              if (index > -1) {
                parent[property].splice(index, 1, reqStatement.statement);
                if (options.onRequireCallExpression) {
                  options.onRequireCallExpression(ImportType.FROM, reqStatement.reqStatement);
                }
              }
            } else {
              const index = parent[property].indexOf(node);
              if (index > -1) parent[property].splice(index, 1);
            }
          },
        };
      }
    },
  };
}
