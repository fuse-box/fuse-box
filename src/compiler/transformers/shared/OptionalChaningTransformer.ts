import { ITransformerSharedOptions } from '../../interfaces/ITransformerSharedOptions';
import { IVisit, IVisitorMod } from '../../Visitor/Visitor';
import { ASTNode } from '../../interfaces/AST';
import { GlobalContext } from '../../program/GlobalContext';

function buildChainTree(node: ASTNode, list: Array<ASTNode>) {
  // if (node.object.type === 'OptionalExpression') {
  //   if (node.object.object.type === 'Identifier') {
  //     list.unshift(node.object.object);
  //   }
  // }
  if (node.type === 'MemberExpression') {
    list.unshift(node);
    return list;
  }
  if (node.object.type === 'Identifier') {
    
    list.unshift(node);
    return list;
  } else {
    list.unshift(node);
    buildChainTree(node.object, list);
  }
}

function moveChain(chain: ASTNode, expression?: ASTNode): ASTNode {
  if (!chain.base) {
    return chain.property;
  }
  if (!expression) expression = { type: 'MemberExpression' };
  expression.property = chain.property;
  expression.computed = chain.computed;
  expression.object = moveChain(chain.base);

  return expression;
}

function conver2MemberExpression(node: ASTNode) {
  if (node.chain) {
    let chain = node.chain;
    return moveChain(chain);
  }

  return node;
}

function createConditionalExpression(sysName: string) {
  // prevMember.test.left.left.right
  // prevMement.alternate.property
  const node: ASTNode = {
    type: 'ConditionalExpression',
    test: {
      type: 'LogicalExpression',
      left: {
        type: 'BinaryExpression',
        left: {
          type: 'AssignmentExpression',
          left: {
            type: 'Identifier',
            name: sysName,
          },
          operator: '=',
          right: null,
        },
        right: {
          type: 'Literal',
          value: null,
        },
        operator: '===',
      },
      right: {
        type: 'BinaryExpression',
        left: {
          type: 'Identifier',
          name: sysName,
        },
        right: {
          type: 'UnaryExpression',
          operator: 'void',
          argument: {
            type: 'Literal',
            value: 0,
          },
          prefix: true,
        },
        operator: '===',
      },
      operator: '||',
    },
    consequent: {
      type: 'UnaryExpression',
      operator: 'void',
      argument: {
        type: 'Literal',
        value: 0,
      },
      prefix: true,
    },
    alternate: {
      type: 'MemberExpression',
      object: {
        type: 'Identifier',
        name: sysName,
      },
      computed: false,
      property: null,
    },
  };
  return node;
}

export function OptionalChaningTransformer(options?: ITransformerSharedOptions) {
  options = options || {};
  return (visit: IVisit): IVisitorMod => {
    const node = visit.node;

    if (node.type === 'OptionalExpression') {
      const globalContext = visit.globalContext as GlobalContext;
      const list: Array<ASTNode> = [];

      buildChainTree(node, list);

      let length = list.length;
      let index = 0;
      let prevMember: ASTNode;

      let childExpression: ASTNode;
      while (index < length) {
        const item = list[index];
        if (item.object && item.object.type === 'Identifier' && item.type !== 'MemberExpression') {
          // this is the first one
          const nextVar = globalContext.getNextSystemVariable();
          prevMember = createConditionalExpression(nextVar);
          prevMember.test.left.left.right = item.object;
          childExpression = prevMember;
        }

        const member = conver2MemberExpression(item);
        if (!childExpression) {
          const nextVar = globalContext.getNextSystemVariable();
          prevMember = createConditionalExpression(nextVar);
          prevMember.test.left.left.right = member;
          childExpression = prevMember;
        } else {
          if (prevMember) {
            prevMember.alternate.property = member;
            prevMember = undefined;
          } else {
            const nextVar = globalContext.getNextSystemVariable();
            const newExpression = createConditionalExpression(nextVar);
            newExpression.test.left.left.right = childExpression;
            newExpression.alternate.property = member;
            console.log('member', member, item);
            childExpression = newExpression;
          }
        }

        index++;
      }

      return { ignoreChildren: true, replaceWith: childExpression };
    }
    return;
  };
}
