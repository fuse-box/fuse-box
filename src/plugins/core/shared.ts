import * as postcss from 'postcss';
import { IStyleSheetProps } from '../../config/IStylesheetProps';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { cssDevModuleRender } from '../../stylesheet/cssDevModuleRender';
import { IStyleSheetProcessor } from '../../stylesheet/interfaces';
import { isNodeModuleInstalled } from '../../utils/utils';
import { IPluginCommon } from '../interfaces';
import { wrapContents } from '../pluginStrings';

export interface ICSSContextHandler {
  ctx: Context;
  module: IModule;
  options: IStyleSheetProps;
  processor: IStyleSheetProcessor;
  shared: IPluginCommon;
}
export function setEmpty() {}

export interface ICreateCSSModule {
  css?: string;
  module: IModule;
  shared: IPluginCommon;
}
async function createCSSModule(props: ICreateCSSModule): Promise<{ json: any; css: string; map: any }> {
  let targetJSON: any;
  return new Promise((resolve, reject) => {
    postcss([
      require('postcss-modules')({
        getJSON: function(cssFileName, json, outputFileName) {
          targetJSON = json;
        },
      }),
    ])
      .process(props.css, {
        from: props.module.absPath,
        map: props.module.isCSSSourceMapRequired && { inline: false },
        to: props.module.absPath,
      })
      .then(result => {
        return resolve({ css: result.css, json: targetJSON, map: result.map });
      });
  });
}
export function cssContextHandler(props: ICSSContextHandler) {
  const { ctx, processor, shared } = props;
  const supported = props.ctx.config.supportsStylesheet() || shared.asText;
  if (!supported) {
    props.module.contents = 'module.exports = {}';
    return;
  }
  if (shared.asModule) {
    if (!isNodeModuleInstalled('postcss-modules')) {
      ctx.fatal(`Fatal error when capturing ${props.module.absPath}`, [
        'Module "postcss-modules" is required, Please install it using the following command',
        'npm install postcss-modules --save-dev',
      ]);
      return;
    }
  }
  ctx.ict.promise(async () => {
    try {
      // reset errored status
      props.module.errored = false;
      const data = await processor.render();
      const rendererProps = {
        ctx,
        data,
        module: props.module,
        options: props.options,
        useDefault: props.shared.useDefault,
      };

      if (shared.asModule) {
        const result = await createCSSModule({ css: data.css, module: props.module, shared: props.shared });
        data.json = result.json;
        data.map = undefined;
        data.css = result.css;
        props.module.isCSSModule = true;
        props.module.breakDependantsCache = true;
      } else if (shared.asText) {
        props.module.isCSSText = true;
        props.module.isStylesheet = false;

        props.module.contents = wrapContents(JSON.stringify(data.css), props.shared.useDefault);

        return;
      }
      if (ctx.config.isProduction) {
        props.module.css = data;
        if (shared.asModule && data.json) {
          props.module.contents = wrapContents(JSON.stringify(data.json), props.shared.useDefault);
        }
      } else {
        cssDevModuleRender({ ...rendererProps });
      }
    } catch (e) {
      // prevent module from being cached
      props.module.errored = true;
      let errMessage = e.message ? e.message : `Uknown error in file ${props.module.absPath}`;
      props.module.contents = `console.error(${JSON.stringify(errMessage)})`;
      ctx.log.error('$error in $file', {
        error: e.message,
        file: props.module.absPath,
      });
    }
  });
}
