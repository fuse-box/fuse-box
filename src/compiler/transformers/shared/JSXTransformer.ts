import { ISchema } from '../../core/nodeSchema';
import { ASTNode, ASTType } from '../../interfaces/AST';
import { ITransformer } from '../../interfaces/ITransformer';

//const Factories: { [key: string]: () => ASTNode } = {};

function createJSXFactory(props: { first: string; second?: string }): (args: Array<ASTNode>) => ASTNode {
  if (!props.second) {
    return (args: Array<ASTNode>): ASTNode => {
      return {
        arguments: args,
        callee: {
          name: props.first,
          type: 'Identifier',
        },
        type: 'CallExpression',
      };
    };
  }
  return (args: Array<ASTNode>): ASTNode => {
    return {
      arguments: args,
      callee: {
        computed: false,
        object: {
          name: props.first,
          type: 'Identifier',
        },
        property: {
          name: props.second,
          type: 'Identifier',
        },
        type: 'MemberExpression',
      },
      type: 'CallExpression',
    };
  };
}

function createObjectAssignExpression(): ASTNode {
  return {
    arguments: [],
    callee: {
      computed: false,
      object: {
        name: 'Object',
        type: 'Identifier',
      },
      property: {
        name: 'assign',
        type: 'Identifier',
      },
      type: 'MemberExpression',
    },
    type: 'CallExpression',
  };
}

export interface IJSXTranformerOptions {
  jsxFactory?: string;
}

function parseFactory(factory: string) {
  const [first, second] = factory.split('.');
  const JSXFragment: ASTNode = {
    computed: false,
    object: {
      name: first,
      type: 'Identifier',
    },
    property: {
      name: 'Fragment',
      type: 'Identifier',
    },
    type: 'MemberExpression',
  };

  const createElement = createJSXFactory({ first, second });
  return { JSXFragment, createElement };
}

function convertJSXMemberExpression(expression: ASTNode) {
  if (expression.type === ASTType.JSXMemberExpression) expression.type = 'MemberExpression';

  if (expression.type === ASTType.JSXIdentifier) expression.type = 'Identifier';

  if (expression.property && expression.property.type === ASTType.JSXIdentifier)
    expression.property.type = 'Identifier';

  if (expression.object) {
    if (expression.object.type === ASTType.JSXMemberExpression) {
      return convertJSXMemberExpression(expression.object);
    }
    if (expression.object.type === ASTType.JSXIdentifier) expression.object.type = 'Identifier';
  }
}

export function JSXTransformer(): ITransformer {
  return {
    commonVisitors: props => {
      const {
        transformationContext: {
          compilerOptions: { jsxFactory },
          module: { extension },
        },
      } = props;

      // we don't need this for normal TypeScript files
      if (extension === '.ts') return;

      // prepare for setting up the jsxFactory
      let initJsxFactory = false;
      let JSXFragment;
      let createElement;

      return {
        onEach: (schema: ISchema) => {
          const { context, node, replace } = schema;
          const name = node.name as string;

          // We only want to setup the jsxFacory once for this module
          if (!initJsxFactory) {
            const factory = context.jsxFactory;

            ({ JSXFragment, createElement } = parseFactory(factory || jsxFactory));
            initJsxFactory = true;
          }

          switch (node.type) {
            case 'JSXElement':
              let props: ASTNode;
              let propObjects: Array<ASTNode> = [];
              let propObject: ASTNode;
              let newObj = true;
              let spreaded = false;

              const { openingElement } = node;

              for (const attr of openingElement.attributes) {
                // less member access
                let { type, value } = attr; // call 'attr' once

                if (type === 'JSXAttribute') {
                  if (!value) {
                    value = { type: 'Literal', value: true };
                  }
                  let key: ASTNode;
                  if (attr.name.name.indexOf('-') > -1) {
                    key = { type: 'Literal', value: attr.name.name };
                  } else {
                    key = { name: attr.name.name, type: 'Identifier' };
                  }

                  const createdProp: ASTNode = {
                    computed: false,
                    key: key,
                    kind: 'init',
                    method: false,
                    shorthand: false,
                    type: 'Property',
                    value: value,
                  };

                  if (newObj) {
                    propObject = {
                      properties: [createdProp],
                      type: 'ObjectExpression',
                    };
                    newObj = false;
                  } else {
                    propObject.properties.push(createdProp);
                  }
                }

                if (type === 'JSXSpreadAttribute') {
                  spreaded = true;
                  if (propObject) {
                    propObjects.push(propObject);
                    // reset for attributes after spread operator
                    propObject = undefined;
                  } else {
                    propObjects.push({
                      properties: [],
                      type: 'ObjectExpression',
                    });
                  }
                  newObj = true;
                  propObjects.push(attr.argument);
                }
              }

              if (spreaded) {
                props = createObjectAssignExpression();
                if (propObject) {
                  propObjects.push(propObject);
                }

                props.arguments = propObjects;
              } else if (propObject) {
                props = propObject;
              } else props = { type: 'Literal', value: null };

              return replace(createElement([openingElement.name, props].concat(node.children)));

            case 'JSXExpressionContainer':
            case 'JSXSpreadChild':
              if (node.expression.type === 'JSXEmptyExpression') return schema.remove();
              return replace(node.expression);
            case 'JSXFragment':
              return replace(
                createElement([JSXFragment, { type: 'Literal', value: null } as ASTNode].concat(node.children)),
              );

            case 'JSXIdentifier':
              if (name[0] === name[0].toLowerCase()) {
                return replace({ type: 'Literal', value: name });
              }
              node.type = 'Identifier';
              return replace(node);
            case 'JSXMemberExpression':
              convertJSXMemberExpression(node);
              return replace(node);
            case 'JSXText':
              if (node.value.indexOf('\n') > -1 && !node.value.trim()) {
                return schema.remove();
              }
              return replace({ type: 'Literal', value: node.value });
          }
        },
      };
    },
  };
}
