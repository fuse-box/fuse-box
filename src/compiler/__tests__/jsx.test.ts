import { initCommonTransform } from '../testUtils';
import { ImportTransformer } from '../transformers/shared/ImportTransformer';
import { JSXTransformer } from '../transformers/shared/JSXTransformer';

const testTranspile = (props: { code: string; jsx?: boolean }) => {
  return initCommonTransform({
    jsx: true,
    transformers: [JSXTransformer(), ImportTransformer()],
    code: props.code,
    props: { module: { props: { extension: '.tsx' } } },
  });
};

describe('JSX', () => {
  it('should parse simple jsx', () => {
    const result = testTranspile({
      code: `
        function test(){
            return <div></div>
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add text', () => {
    const result = testTranspile({
      code: `
        function test(){
            return <div>1</div>
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add just attributes', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id="1" f={oi} ></i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should add key literal attributes', () => {
    const result = testTranspile({
      code: `
      function hey() {
        return <div data-automation="df"></div>;
      }`,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should convert attributes without vales to boolean', () => {
    const result = testTranspile({
      code: `
      function hey() {
        return <div some></div>;
      }`,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should add just attributes and spread', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id="1" f={oi} {...props} ></i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should have only one spread props', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i {...props} ></i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });

  it('should keep the order', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i id={1} f={oi} {...props} s={2} {...rest}></i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should remove empty expressions', () => {
    const result = testTranspile({
      code: `
        import oi from "./oi";
        function test(){
          return (<i> {/* <i></i> */} </i>)
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should handle JSX fragment', () => {
    const result = testTranspile({
      code: `
        import React from "react";
        function test(){
          return (
            <div>
              <>
                <div>1</div>
              </>
            </div>
          )
        }
          `,
    });

    expect(result.code).toMatchSnapshot();
  });

  it('should work with JSXSpreadChild', () => {
    const result = testTranspile({
      code: `

        function test(){
          return (<i>{...children}</i>)
        }
          `,
    });
    expect(result.code).toMatchSnapshot();
  });
});
