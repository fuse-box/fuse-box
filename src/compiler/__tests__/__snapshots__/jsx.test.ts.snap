// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX should add just attributes 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", {
    id: \\"1\\",
    f: oi_1.default
  });
}
"
`;

exports[`JSX should add just attributes and spread 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", Object.assign({
    id: \\"1\\",
    f: oi_1.default
  }, props));
}
"
`;

exports[`JSX should add key literal attributes 1`] = `
"function hey() {
  return React.createElement(\\"div\\", {
    \\"data-automation\\": \\"df\\"
  });
}
"
`;

exports[`JSX should add text 1`] = `
"function test() {
  return React.createElement(\\"div\\", null, \\"1\\");
}
"
`;

exports[`JSX should convert attributes without vales to boolean 1`] = `
"function hey() {
  return React.createElement(\\"div\\", {
    some: true
  });
}
"
`;

exports[`JSX should handle JSX fragment 1`] = `
"var react_1 = require(\\"react\\");
function test() {
  return react_1.default.createElement(\\"div\\", null, react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement(\\"div\\", null, \\"1\\")));
}
"
`;

exports[`JSX should handle dotted JSX elements 1`] = `
"var react_1 = require(\\"react\\");
function test() {
  return react_1.default.createElement(animated.div, null, \\"1\\");
}
"
`;

exports[`JSX should have only one spread props 1`] = `
"function test() {
  return React.createElement(\\"i\\", Object.assign({}, props));
}
"
`;

exports[`JSX should keep the order 1`] = `
"var oi_1 = require(\\"./oi\\");
function test() {
  return React.createElement(\\"i\\", Object.assign({
    id: 1,
    f: oi_1.default
  }, props, {
    s: 2
  }, rest));
}
"
`;

exports[`JSX should parse simple jsx 1`] = `
"function test() {
  return React.createElement(\\"div\\", null);
}
"
`;

exports[`JSX should remove empty expressions 1`] = `
"function test() {
  return React.createElement(\\"i\\", null, \\" \\", \\" \\");
}
"
`;

exports[`JSX should spread correctly 1`] = `
"function EventList() {
  const cameraIDs = ['a'];
  return React.createElement(Foo, Object.assign({
    jo: 'test'
  }, entry, {
    cameraIDs: cameraIDs,
    key: '123'
  }));
}
"
`;

exports[`JSX should work with JSXSpreadChild 1`] = `
"function test() {
  return React.createElement(\\"i\\", null, children);
}
"
`;
